test/src/main/java/org/apache/accumulo/test/functional/FlushNoFileIT.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Mutation;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.IteratorUtil;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.harness.AccumuloClusterHarness;
+import org.apache.hadoop.io.Text;
+import org.junit.Test;
+
+/**
+ * Tests that Accumulo will flush but not create a file that has 0 entries.
+ */
+public class FlushNoFileIT extends AccumuloClusterHarness {
+
+  public static class NullIterator implements SortedKeyValueIterator<Key,Value> {
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options,
+        IteratorEnvironment env) {}
+
+    @Override
+    public boolean hasTop() {
+      return false;
+    }
+
+    @Override
+    public void next() throws IOException {}
+
+    @Override
+    public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) {}
+
+    @Override
+    public Key getTopKey() {
+      return null;
+    }
+
+    @Override
+    public Value getTopValue() {
+      return null;
+    }
+
+    @Override
+    public SortedKeyValueIterator<Key,Value> deepCopy(IteratorEnvironment env) {
+      return null;
+    }
+  }
+
+  @Override
+  protected int defaultTimeoutSeconds() {
+    return 60;
+  }
+
+  @Test
+  public void test() throws Exception {
+    try (AccumuloClient c = Accumulo.newClient().from(getClientProps()).build()) {
+
+      String tableName = getUniqueNames(1)[0];
+
+      NewTableConfiguration ntc = new NewTableConfiguration();
+      IteratorSetting iteratorSetting = new IteratorSetting(20, NullIterator.class);
+      ntc.attachIterator(iteratorSetting, EnumSet.of(IteratorUtil.IteratorScope.minc));
+
+      c.tableOperations().create(tableName, ntc);
+
+      try (BatchWriter bw = c.createBatchWriter(tableName)) {
+        Mutation m = new Mutation(new Text("r1"));
+        m.put("acf", tableName, "1");
+        bw.addMutation(m);
+      }
+
+      FunctionalTestUtils.checkRFiles(c, tableName, 1, 1, 0, 0);
+
+      c.tableOperations().flush(tableName, null, null, true);
+
+      FunctionalTestUtils.checkRFiles(c, tableName, 1, 1, 0, 0);
+
+      long flushId = FunctionalTestUtils.checkFlushId(c, tableName, 0);
+
+      try (BatchWriter bw = c.createBatchWriter(tableName)) {
+        Mutation m = new Mutation(new Text("r2"));
+        m.put("acf", tableName, "1");
+        bw.addMutation(m);
+      }
+
+      c.tableOperations().flush(tableName, null, null, true);
+
+      FunctionalTestUtils.checkRFiles(c, tableName, 1, 1, 0, 0);
+
+      long secondFlushId = FunctionalTestUtils.checkFlushId(c, tableName, flushId);
+      assertTrue("Flush ID did not change", secondFlushId > flushId);
+    }
```

Comment:
Would also be nice to do a scan of the table.   Just to ensure the scan does not throw any errors for this case.